rfcs/text/0010_service_status.md

```diff
@@ -0,0 +1,393 @@
+  /** server.uuid */
+  uuid: string;
+  /** Currently exposed by existing status API */
+  version: {
+    number: string;
+    build_hash: string;
+    build_number: number;
+    build_snapshot: boolean;
+  };
+  /** Similar format to existing API, but slightly different shape */
+  status: {
+    /** See "Overall status calculation" section below */
+    overall: ServiceStatus;
+    core: CoreStatus;
+    plugins: Record<string, ServiceStatus>;
+  }
+}
+```
+
+## Behaviors
+
+### Levels
+
+Each member of the `ServiceStatusLevel` enum has specific behaviors associated with it:
+- **`available`**:
+  - All endpoints and apps associated with the service are accessible
+- **`degraded`**:
+  - All endpoints and apps are available by default
+  - Some APIs may return `503 Unavailable` responses. This is not automatic, must be implemented directly by the service.
+  - Some plugin contract APIs may throw errors. This is not automatic, must be implemented directly by the service.
+- **`unavailable`**:
+  - All endpoints (with some exceptions in Core) in Kibana return a `503 Unavailable` responses by default. This is automatic.
+  - When trying to access any app associated with the unavailable service, the user is presented with an error UI with detail about the outage.
+  - Some plugin contract APIs may throw errors. This is not automatic, must be implemented directly by the service.
+- **`critical`**:
+  - All endpoints (with some exceptions in Core) in Kibana return a `503 Unavailable` response by default. This is automatic.
+  - All applications redirect to the system-wide status page with detail about which services are down and any relevant detail. This is automatic.
+  - Some plugin contract APIs may throw errors. This is not automatic, must be implemented directly by the service.
+  - This level is reserved for Core services only.
+
+### Overall status calculation
+
+The status level of the overall system is calculated to be the highest severity status of all core services and plugins.
+
+The `summary` property is calculated as follows:
+- If the overall status level is `available`, the `summary` is `"Kibana is operating normally"`
+- If a single core service or plugin is not `available`, the `summary` is `Kibana is ${level} due to ${serviceName}. See ${statusPageUrl} for more information.`
+- If multiple core services or plugins are not `available`, the `summary` is `Kibana is ${level} due to multiple components. See ${statusPageUrl} for more information.`
+
+### Status inheritance
+
+By default, plugins inherit their status from all Core services and their dependencies on other plugins. The default status for a plugin follows this algorithm:
+```js
+// pseudo-code
+function getPluginStatus() {
+  if any core service is `critical`, return `critical`
+  else if any core service is `unavailable`, return `unavailable`
+  else if any core service is `degraded`, return `degraded`
+
+  if any required dep is `unavailable`, return `unavailable`
```

Comment:
'any required dep is `unavailable`' + 'any core service is `degraded`' returns `degraded` here, is that wanted?

Reply:
Fixed